integer overflow

I recently ran into the following panic when attempting to open a large parquet file. 

```
panic: runtime error: index out of range [-425984]

goroutine 1 [running]:
github.com/segmentio/parquet-go.(*fileRowGroup).init(0x1401ebc6000, 0x14000182360, 0x10282d040?, {0x14000147880, 0xd, 0x1027adee0?}, 0x14002ec0000)
        /Users/thor/go/src/github.com/polarsignals/frostdb/vendor/github.com/segmentio/parquet-go/file.go:381 +0x4b0
github.com/segmentio/parquet-go.OpenFile({0x102828d98?, 0x1400012a050}, 0x37926e50, {0x0, 0x0, 0x0})
        /Users/thor/go/src/github.com/polarsignals/frostdb/vendor/github.com/segmentio/parquet-go/file.go:100 +0x760
```

This is caused due to the fact that the writer will cast a length without checking for integer overflow. https://github.com/segmentio/parquet-go/blob/771b3e358a03c412947573a15fe7af1e1eb3751a/writer.go#L640 

Looking through the code a bit more I noticed that this problem exists in a lot of places

ex: 
https://github.com/segmentio/parquet-go/blob/771b3e358a03c412947573a15fe7af1e1eb3751a/writer.go#L606

https://github.com/segmentio/parquet-go/blob/771b3e358a03c412947573a15fe7af1e1eb3751a/writer.go#L333

While most of these are likely much harder to hit since they are wider integers the possibility still exists to hit this bug in other parts of the code. 

I suggest we at least convert the `Ordinal` value to a `uint16` since it being negative has no meaning, as well as add some casting guards in the writer codes to catch overflow at write time. Further looking it may be a good idea to convert all of these ints into their uint counterparts since there doesn't seem to be value to being able to represent negative numbers from what I can tell.

--- a/errors.go
+++ b/errors.go
@@ -49,6 +49,10 @@ var (
 	// ErrUnexpectedDefinitionLevels is an error returned when attempting to
 	// decode definition levels into a page which is part of a required column.
 	ErrUnexpectedDefinitionLevels = errors.New("unexpected definition levels")
+
+	// ErrTooManyRowGroups is returned when attempting to generate a parquet
+	// file with more than MaxRowGroups row groups.
+	ErrTooManyRowGroups = errors.New("the limit of 65535 row groups has been reached")
 )
 
 type errno int

--- a/limits.go
+++ b/limits.go
@@ -12,11 +12,21 @@ const (
 	// MaxColumnIndex is the maximum column index supported by this package.
 	MaxColumnIndex = math.MaxInt16
 
-	// MaxRepetitionLevel is the maximum repetition level supported by this package.
+	// MaxRepetitionLevel is the maximum repetition level supported by this
+	// package.
 	MaxRepetitionLevel = math.MaxUint8
 
-	// MaxDefinitionLevel is the maximum definition level supported by this package.
+	// MaxDefinitionLevel is the maximum definition level supported by this
+	// package.
 	MaxDefinitionLevel = math.MaxUint8
+
+	// MaxRowGroups is the maximum number of row groups which can be contained
+	// in a single parquet file.
+	//
+	// This limit is enforced by the use of 16 bits signed integers in the file
+	// metadata footer of parquet files. It is part of the parquet specification
+	// and therefore cannot be changed.
+	MaxRowGroups = math.MaxInt16
 )
 
 func makeRepetitionLevel(i int) byte {

--- a/writer.go
+++ b/writer.go
@@ -105,6 +105,9 @@ func (w *Writer) Close() error {
 // Flush is called automatically on Close, it is only useful to call explicitly
 // if the application needs to limit the size of row groups or wants to produce
 // multiple row groups per file.
+//
+// If the writer attempts to create more than MaxRowGroups row groups the method
+// returns ErrTooManyRowGroups.
 func (w *Writer) Flush() error {
 	if w.writer != nil {
 		return w.writer.flush()
@@ -540,6 +543,10 @@ func (w *writer) writeRowGroup(rowGroupSchema *Schema, rowGroupSortingColumns []
 	numRows := w.columns[0].totalRowCount()
 	if numRows == 0 {
 		return 0, nil
+	}
+
+	if len(w.rowGroups) == MaxRowGroups {
+		return 0, ErrTooManyRowGroups
 	}
 
 	defer func() {

